XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+// to the handles themselves. Thus possibly preventing GC from cleaning up
+// the handle and ever firing the destroy() callback.
+const handle = async_wrap.getHandleById(n);
+
+// init() is called during object construction. The handle will have not
+// completed construction when this callback runs. So not all fields will have
+// populated.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed.
+function destroy(id) { }
+
+// Create a new instance and add each callback to the corresponding hook.
+var asyncHook = new AsyncHook({ init, pre, post, destroy });
+
+// Enable hooks for the current synchronous execution scope. This will ensure
+// the hooks are not in effect in case of multiple returns, or if an exception
+// is thrown.
+asyncHook.scope();
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+// Unlike disable(), prevent any hook callback from firing again in the future.
+// Future hooks can be added by running scope()/enable() again.
+// TODO: I am not sure all of this is possible. Some investigation is necessary.
+asyncHook.remove();
+```
+
+
+### `async_wrap`
+
+The object returned from `require('async_wrap')`.
+
+
+#### `async_wrap.subsystems`
+
+List of all subsystems that may trigger the `init` callback. These subsystems
+group relevant calls together. For example the `CRYPTO` provider encompasses
+all calls from the crypto library that only creates a request. Purposes of this
+is to allow quick filtering or reporting of incoming requests.
+
+**Note(trevnorris):** "provider" was originally intended to be "subsystem" then
+later essentially turned into the class name, unless the class being
+instantiated inherited directly from `AsyncWrap`. so should subsystems simply be
+changed to include the name of every class constructor? or should it properly
+group calls together, such as `TCPWRAP` and `TCPCONNECTWRAP` into a single TCP
+provider? or possibly both.
+
```

Comment:
at this stage, without further feedback, you should just remove this paragraph and assert one or the other. "subsystem" seems good enough to me but neither option seems more clear than the other so just go with what you think is appropriate since you've been poking at this for so long and have a better feel for the terminology that makes sense. If you go with "subsystem", remove the remaining instance of "provider" in the doc (para above).